File: 260220/최고의 33위치/best-place-of-33.cpp
#include <cstdio>
#include <cmath>
using namespace std;

constexpr int MAX_BOARD_SIZE = 20;
constexpr int SUB_BOARD_SIZE = 3;

int answer;
int board_size;
int board[MAX_BOARD_SIZE][MAX_BOARD_SIZE];

void input() {
    scanf("%d", &board_size);

    for (int y = 0; y < board_size; ++y) {
        for (int x = 0; x < board_size; ++x) {
            scanf("%d", &board[y][x]);
        }
    }
}

void solve() {
    answer = 0;
    int current;

    for (int y = 0; y < board_size - SUB_BOARD_SIZE + 1; ++y) {
        for (int x = 0; x < board_size - SUB_BOARD_SIZE + 1; ++x) {

            current = 0;

            for (int cy = 0; cy < SUB_BOARD_SIZE; ++cy) {
                for (int cx = 0; cx < SUB_BOARD_SIZE; ++cx) {
                    current += board[y + cy][x + cx];
                }
            }

            answer = max(answer, current);

        }
    }
}

void print() {
    printf("%d\n", answer);
}

int main() {
    setbuf(stdout, NULL);

    input();
    solve();
    print();

    return 0;
}

File: 260714/최고의 33위치/best-place-of-33.java
import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
    static StringBuilder output = new StringBuilder();
    static StringTokenizer input;

    public static void main(String[] args) throws Exception {
        new Main().run();
        System.out.println(output.toString());
    }

    int answer;
    int boardSize;
    char[][] board;

    public Main() {}
    public void run() throws Exception { input(); solve(); print(); }

    private void input() throws IOException {
        boardSize = nextInt();

        board = new char[boardSize][boardSize];

        for (int y = 0; y < boardSize; y++) {
            for (int x = 0; x < boardSize; x++) {
                board[y][x] = next().charAt(0);
            }
        }
    }

    private void solve() {
        answer = 0;

        for (int y = 0; y <= boardSize - 3; y++) {
            for (int x = 0; x <= boardSize - 3; x++) {
                answer = Math.max(answer, cntMat3x3(x, y));
            }
        }
    }

    private int cntMat3x3(int x, int y) {
        int cnt = 0;

        for (int cx = 0; cx < 3; cx++) {
            for (int cy = 0; cy < 3; cy++) {
                if (board[y + cy][x + cx] == '1') cnt++;
            }
        }

        return cnt;
    }

    private void print() {
        output.append(answer);
        output.append('\n');
    }

    String next() throws IOException {
        if (input == null || !input.hasMoreTokens()) {
            input = new StringTokenizer(reader.readLine().trim());
        }
        return input.nextToken();
    }

    int nextInt() throws IOException {
        return Integer.parseInt(next());
    }
}